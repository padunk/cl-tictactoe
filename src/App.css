#root {
  max-width: 1280px;
  margin: 0 auto;
  padding: 2rem;
  text-align: center;
}

.row {
  display: flex;
  justify-content: center;
  align-items: center;
}

.title {
  text-align: center;
  color: orange;
}

.cell {
  border: 1px solid white;
  width: 40px;
  height: 40px;
  color: white;
  display: flex;
  justify-content: center;
  align-items: center;
}
