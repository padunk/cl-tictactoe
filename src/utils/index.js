export const checkWinner = (winningStates, currentBoard) => {
  for (let i = 0; i < winningStates.length; i++) {
    const winningState = winningStates[i];
    const [row0, col0] = winningState[0];
    const [row1, col1] = winningState[1];
    const [row2, col2] = winningState[2];

    if (
      currentBoard[row0][col0] &&
      currentBoard[row0][col0] === currentBoard[row1][col1] &&
      currentBoard[row1][col1] === currentBoard[row2][col2]
    ) {
      return currentBoard[row0][col0];
    }
  }
  return null;
};

export const checkDraw = (currentBoard) => {
  return currentBoard.every((row) => {
    return row.every((r) => Boolean(r));
  });
};
