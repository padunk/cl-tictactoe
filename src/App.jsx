import "./App.css";

import * as React from "react";

import Cell from "./components/cell";
import lo from "lodash";

const board = [
  [null, null, null],
  [null, null, null],
  [null, null, null],
];

// we export this to test it
export const winningStates = [
  [
    [0, 0],
    [0, 1],
    [0, 2],
  ],
  [
    [1, 0],
    [1, 1],
    [1, 2],
  ],
  [
    [2, 0],
    [2, 1],
    [2, 2],
  ],
  [
    [0, 0],
    [1, 0],
    [2, 0],
  ],
  [
    [0, 1],
    [1, 1],
    [2, 1],
  ],
  [
    [0, 2],
    [1, 2],
    [2, 2],
  ],
  [
    [0, 0],
    [1, 1],
    [2, 2],
  ],
  [
    [0, 2],
    [1, 1],
    [2, 0],
  ],
];

const PLAYER_A = "X";
const PLAYER_B = "O";

function App() {
  const [currentPlayer, setNextPlayer] = React.useState("playerA");
  const [currentBoard, setCurrentBoard] = React.useState(board);

  const handleClick = (row, column) => {
    const tempBoardState = lo.cloneDeep(currentBoard);
    if (tempBoardState[row][column] !== null) {
      alert("The cell is already selected. Please choose other cell.");
      return;
    }
    if (tempBoardState[row][column] === null) {
      tempBoardState[row][column] =
        currentPlayer === "playerA" ? PLAYER_A : PLAYER_B;
    }

    setNextPlayer(currentPlayer === "playerA" ? "playerB" : "playerA");
    setCurrentBoard(tempBoardState);
  };

  const checkWinner = () => {
    for (let i = 0; i < winningStates.length; i++) {
      const winningState = winningStates[i];
      const [row0, col0] = winningState[0];
      const [row1, col1] = winningState[1];
      const [row2, col2] = winningState[2];

      if (
        currentBoard[row0][col0] &&
        currentBoard[row0][col0] === currentBoard[row1][col1] &&
        currentBoard[row1][col1] === currentBoard[row2][col2]
      ) {
        let scoreHistory = JSON.parse(localStorage.getItem("tictactoe"));
        if (scoreHistory === null) {
          scoreHistory = { a: 0, b: 0 };
        }
        if (currentBoard[row0][col0] === PLAYER_A) {
          scoreHistory.a++;
        } else {
          scoreHistory.b++;
        }

        localStorage.setItem("tictactoe", JSON.stringify(scoreHistory));
        return currentBoard[row0][col0];
      }
    }
    return null;
  };

  const checkDraw = () => {
    return currentBoard.every((row) => {
      return row.every((r) => Boolean(r));
    });
  };

  const isWinner = checkWinner();
  const isDraw = checkDraw();

  return (
    <div className="App">
      <h1 className="title">TIC TAC TOE</h1>
      <h3>Player A: {PLAYER_A}</h3>
      <h3>Player B: {PLAYER_B}</h3>
      {isDraw ? (
        <>
          <h2 data-testid="test.draw">It's a DRAW!</h2>
          <button onClick={() => setCurrentBoard(board)}>Restart</button>
        </>
      ) : isWinner ? (
        <>
          <h2 data-testid="test.winner">
            Player {isWinner === PLAYER_A ? "A" : "B"} has win
          </h2>
          <button onClick={() => setCurrentBoard(board)}>Restart</button>
        </>
      ) : (
        currentBoard.map((b, idx) => {
          return (
            <div key={idx} className="row">
              <Cell text={b[0]} onClick={() => handleClick(idx, 0)} />
              <Cell text={b[1]} onClick={() => handleClick(idx, 1)} />
              <Cell text={b[2]} onClick={() => handleClick(idx, 2)} />
            </div>
          );
        })
      )}
    </div>
  );
}

export default App;
