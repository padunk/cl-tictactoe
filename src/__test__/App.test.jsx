import App, { winningStates } from "../App";
import { checkDraw, checkWinner } from "../utils";
import { describe, expect, it } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";

import Cell from "../components/cell";
import userEvent from "@testing-library/user-event";

const user = userEvent.setup();

const board = [
  [null, null, null],
  [null, null, null],
  [null, null, null],
];

describe("App", () => {
  it("matching the snapshot", () => {
    render(<App />);

    expect(<App />).toMatchSnapshot();
  });

  it("click the cell will change the inner text", async () => {
    const { getAllByTestId } = render(<App />);
    const buttons = getAllByTestId("test.cell");

    await user.click(buttons[0]);
    expect(buttons[0].innerHTML).toBe("X");

    await user.click(buttons[0]);
    expect(buttons[0].innerHTML).not.toBe("O"); // negative test

    await user.click(buttons[1]);
    expect(buttons[1].innerHTML).toBe("O");
  });

  it("should show winner title", async () => {
    const { getAllByTestId, getByTestId } = render(<App />);
    const buttons = getAllByTestId("test.cell");

    await user.click(buttons[0]); // X
    await user.click(buttons[4]); // O
    await user.click(buttons[1]); // X
    await user.click(buttons[5]); // O
    await user.click(buttons[2]); // X

    expect(getByTestId("test.winner")).toBeInTheDocument();
  });

  it("should show draw title", async () => {
    const { getAllByTestId, getByTestId } = render(<App />);
    const buttons = getAllByTestId("test.cell");

    await user.click(buttons[0]); // X
    await user.click(buttons[1]); // O
    await user.click(buttons[2]); // X
    await user.click(buttons[6]); // O
    await user.click(buttons[7]); // X
    await user.click(buttons[8]); // O
    await user.click(buttons[3]); // X
    await user.click(buttons[4]); // O
    await user.click(buttons[5]); // X

    expect(getByTestId("test.draw")).toBeInTheDocument();
  });

  it("should render correct text", () => {
    const { getByTestId } = render(<Cell text="X" onClick={() => {}} />);

    expect(getByTestId("test.cell").innerHTML).toEqual("X");
  });
});

describe("Unit Test", () => {
  test("should not return winner", () => {
    let result = checkWinner(winningStates, board);
    expect(result).toBeNull();

    board[0] = ["X", "X", "O"];
    result = checkWinner(winningStates, board);
    expect(result).toBeNull();

    board[0] = ["X", "X", "O"];
    board[1] = ["O", "X", "O"];
    board[2] = [null, null, null];

    result = checkWinner(winningStates, board);
    expect(result).toBeNull();
  });

  test("should return a winner", () => {
    let result = checkWinner(winningStates, board);
    expect(result).toBeNull();

    board[0] = ["X", "X", "X"];
    result = checkWinner(winningStates, board);
    expect(result).toEqual("X");

    board[0] = ["X", "X", "O"];
    board[1] = ["O", "O", "O"];
    board[2] = [null, null, null];

    result = checkWinner(winningStates, board);
    expect(result).toEqual("O");

    board[0] = ["X", "X", "O"];
    board[1] = ["O", "X", "O"];
    board[2] = [null, null, "O"];

    result = checkWinner(winningStates, board);
    expect(result).toEqual("O");

    board[0] = ["X", "X", "O"];
    board[1] = ["O", "X", "O"];
    board[2] = [null, null, "X"];

    result = checkWinner(winningStates, board);
    expect(result).toEqual("X");
  });

  test("should be draw", () => {
    let result = checkDraw(board);
    expect(result).toBe(false);

    board[0] = ["X", "O", "X"];
    result = checkDraw(board);
    expect(result).toBe(false);

    board[0] = ["X", "X", "O"];
    board[1] = ["O", "O", "O"];
    board[2] = [null, null, null];

    result = checkDraw(board);
    expect(result).toBe(false);

    board[0] = ["X", "X", "O"];
    board[1] = ["O", "X", "O"];
    board[2] = [null, null, "O"];

    result = checkDraw(board);
    expect(result).toEqual(false);

    board[0] = ["X", "O", "O"];
    board[1] = ["O", "O", "X"];
    board[2] = ["O", "X", "X"];

    result = checkDraw(board);
    expect(result).toEqual(true);
  });
});
