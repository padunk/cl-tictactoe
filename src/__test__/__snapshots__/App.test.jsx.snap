// Vitest Snapshot v1

exports[`App > matching the snapshot 1`] = `<App />`;
